### ISSUE TITLE
Set and Dictionary should have `remove(where:)`.

### ISSUE THREAD

**Issue by dabrahams** [swiftlang/swift#78364]
### Motivation

It's not possible to implement this operation efficiently from outside the library.  You either iterate a snapshot and end up silently copying the collection, or you have to collect the items to be removed in a separate collection and then `.subtract` that.

### Proposed solution

Add `remove(where:)` to `Set` and `Dictionary`.

**Comment by natecook1000** [comment]
Agreed, with the name `removeAll(where:)` to match the [RangeReplaceableCollection requirement](https://developer.apple.com/documentation/swift/rangereplaceablecollection/removeall(where:)).

**Comment by krishpranav** [comment]
Hi, is this still open?

**Comment by krishpranav** [comment]
@ maintainers if possible can you assign it to me so that i will raise an PR for this.

**Comment by AnthonyLatsis** [comment]
@krishpranav Please note that this addition requires a [Swift evolution proposal](https://www.swift.org/swift-evolution) — hence the corresponding label.

**Comment by krishpranav** [comment]
so should i need to create a new proposal like that?

**Comment by AnthonyLatsis** [comment]
You are welcome to! Make sure to follow the process document that is linked there.

**Comment by krishpranav** [comment]
pitch: (https://github.com/swiftlang/swift-evolution/pull/2647)

**Comment by stephentyrone** [comment]
@dabrahams I've been thinking about the polarity of this operation. `Set` already has `.filter { ... }`, so the most obvious spelling here might be `set.filtered { !predicate }`. Would you prefer `remove[All](where:)`, or is that just the name that came to mind?

Also some vague spitballing about how we might implement a set partition efficiently...

**Comment by dabrahams** [comment]
I find `filter` terrible as a name terrible, notwithstanding term-of-art precedent, because it doesn't tell you the polarity of the predicate, and the word even suggests the opposite polarity (we “filter out,” never in) and yes, I do forget from time to time.  See also [this thread](https://forums.swift.org/t/draft-apply-ed-ing-rule-to-core-functional-methods-e-g-filter-filtered/3011/29) (and probably others).  As noted there I would prefer `nonPrimes.where(\.isEven)` for the positive cases.  But as long as there are `remove` methods I think it's valuable to preserve the mental link for the developer who starts with the idea of removing something.  IIRC I objected to the name `removeAll` for these, because “all” doesn't add anything and even suggests a close relationship to `removeAll()`, but I lost that battle (or never fought it) and the ship sailed.

Since you can't reorder the items in a set, it's not clear to me what you might mean by “a set partition.”

**Comment by lorentey** [comment]
We can definitely provide an operation that splits a set/dictionary into two parts based on some provided predicate, and in a way that's better than manually looping through items.

We do not have tombstones, so we cannot generally avoid rehashing items, even with in-place removals (because we need to compress collision chains). But there is still plenty of opportunity for optimization -- for example, for bulk operations it's been particularly helpful to precalculate predicate results into a temporary bitmap in an up front pass before selecting a strategy. (The bitmap gives us a precise picture of the size of the results, so we can avoid repeated resizings of newly allocated hash table results, and we can heuristically switch between in-place removals and building a new table based on the number of items that remain in place. Bulk in-place removals are often considerably slower than copying the remaining elements directly into a brand new, well-sized hash table. In-place removals can also leave the table wildly oversized, which is sometimes unexpected/wasteful.)

**Comment by dabrahams** [comment]
Since your set representation is resilient, it might be possible to leave the underlying storage alone and simply refer to it from something that stores a bitset to tell you which elements are actually present in that partition.  All the set operations would be the same; only the collection index movement operations would change.

**Comment by lorentey** [comment]
The standard `Set`/`Dictionary` types have mostly frozen representations -- they do not have resilient storage. (Although the hash function is opaque, and we have intentionally left a few storage aspects out of the ABI. E.g., the maximum load factor is one knob we can still tweak; but the lookup algorithm is mostly etched in stone.) 

(The generic lookup paths needed to remain fully specializable, to avoid shipping a dramatic performance regression in Swift 5.)

**Comment by krishpranav** [comment]
Thank you all for the insightful discussion and feedback! 🙏  

### Regarding Naming:
I agree that preserving clarity in predicate polarity is essential. While `removeAll(where:)` may already be familiar, I understand the concerns about "all" potentially being misleading. I lean towards keeping `remove(where:)` as it directly conveys intent without implying exhaustive removal semantics tied to `removeAll()`.  

Alternatively, introducing a positive-predicate counterpart could provide balance—perhaps a `retain(where:)` function to complement `remove(where:)`.  

### On Partitioning:
The idea of introducing an efficient `partition(by:)` operation is compelling. It aligns with providing an optimized way to split collections without requiring manual iteration. While we can’t avoid rehashing in most cases, I like the proposed strategy of leveraging a bitmap for performance gains. Precomputing predicate results to optimize hash table resizing or selectively choosing between in-place removal vs. reconstruction is a smart approach.  

### Follow-Up:
Given the constraints of frozen representations in the standard `Set`/`Dictionary`, I wonder if introducing a `partition(by:)` function with a shared bitmap-backed structure (for both partitions) would balance performance with clarity. For `remove(where:)`, I’m ready to finalize the implementation and accompanying tests unless there’s a desire to bundle it with a partition-related feature for broader cohesion.

Would you recommend splitting this work into separate PRs or considering a proposal to bundle the related functionality?  

Thanks again for your guidance!

**Comment by dabrahams** [comment]
IMO (sadly) if you're going with `remove(where:)` instead of `removeAll(where:)` you need to deprecate/rename the other instances of that suboptimal name in the standard library.
